File: angular/src/app/core/services/calendar.service.ts
import { inject, Injectable } from '@angular/core';
import {
  Firestore,
  collection,
  collectionData,
  addDoc,
  deleteDoc,
  doc,
  query,
  where,
  serverTimestamp,
} from '@angular/fire/firestore';
import { Observable } from 'rxjs';
import { CalendarEvent } from '../models/calendar-event.model';

@Injectable({ providedIn: 'root' })
export class CalendarService {
  private firestore = inject(Firestore);

  /**
   * Événements auxquels l'utilisateur participe ou qu'il a créés
   */
  getEvents$(userId: string): Observable<CalendarEvent[]> {
    const q = query(
      collection(this.firestore, 'calendar_events'),
      where('participantsIds', 'array-contains', userId)
    );
    return collectionData(q, { idField: 'id' }) as Observable<CalendarEvent[]>;
  }

  createEvent(event: Omit<CalendarEvent, 'id'>): Promise<void> {
    const data = Object.fromEntries(
      Object.entries({ ...event, createdAt: serverTimestamp() })
        .filter(([, v]) => v !== undefined)
    );
    return addDoc(collection(this.firestore, 'calendar_events'), data).then(() => undefined);
  }

  deleteEvent(eventId: string): Promise<void> {
    return deleteDoc(doc(this.firestore, `calendar_events/${eventId}`));
  }
}

File: flutter/lib/core/services/calendar_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/calendar_event.dart';

/// Service de gestion du calendrier — miroir de CalendarService Angular.
class CalendarService {
  CalendarService({FirebaseFirestore? firestore})
      : _db = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _db;

  // ─── Lecture temps réel ───────────────────────────────────────────────────

  /// Événements auxquels l'utilisateur participe.
  Stream<List<CalendarEvent>> getEvents(String userId) {
    return _db
        .collection('calendar_events')
        .where('participantsIds', arrayContains: userId)
        .snapshots()
        .map((snap) {
      final list = snap.docs
          .map((d) => CalendarEvent.fromDoc(
              d as DocumentSnapshot<Map<String, dynamic>>))
          .toList();
      list.sort((a, b) => a.debut.compareTo(b.debut));
      return list;
    });
  }

  // ─── Écriture ─────────────────────────────────────────────────────────────

  Future<void> createEvent(CalendarEvent event) async {
    await _db.collection('calendar_events').add({
      ...event.toMap(),
      'createdAt': FieldValue.serverTimestamp(),
    });
  }

  Future<void> deleteEvent(String eventId) async {
    await _db.collection('calendar_events').doc(eventId).delete();
  }
}
